### PR TITLE
DEVPROD-19209: Core Log Summarizer Workflow

### PR DESCRIPTION
**Pull request by Julien-Ben**
# Advanced Log Analysis Workflow with Iterative Refinement

[DEVPROD-19209](https://jira.mongodb.org/browse/DEVPROD-19209)

## Summary

This PR introduces the **Log Core Analyzer Workflow**, enabling the Parsley Agent to analyze technical files of any format and size. The workflow automatically adapts its processing strategy based on file size: using single-pass analysis for small files and iterative refinement for large ones.

## Example run

```
yarn run-analyzer --batch test_data/
🔬 Analyzing 5 files

[...]

════════════════════════════════════════════════════════════
📊 ANALYSIS SUMMARY
────────────────────────────────────────────────────────────

📈 Overall Statistics:
   • Files processed: 5
   • Total size: 1.3MB
   • Success rate: 5/5 (100.0%)
   • Total time: 107.7s
   • Processing speed: 12.6KB/s

⏱️  Performance Metrics:
   • Average time: 21.5s
   • Average size: 270.9KB
   • Slowest: easy_sharded_cluster_custom_podspec_max_api_keys (30.7s, 504.7KB)
   • Fastest: easy_python_tests_build_info_fail (15.4s, 92.8KB)
   • Largest: easy_svet_compile_failed (605.1KB)

💾 Detailed results saved to: tmp/test-results-2025-08-26T17-36-02.json

📋 Results:
────────────────────────────────────────────────────────────

  ✅ easy_build_agent_images_ubi_missing_function
     Status: Success
     Size: 61.2KB | Time: 16.8s
     Summary: The build process for a Python project failed due to an AttributeError in the `BuildConfiguration` class, specifically a missing `is_release_step_executed` attribute. Despite successful installation of 12 packages and two build attempts, the error persisted, preventing completion. Immediate action is required to investigate and restore the missing attribute in the build script, followed by a re-run to confirm resolution. Compatibility of environment setup and package versions should also be validated.
     Report: tmp/reports/report-2025-08-26T17-34-31.md

  ✅ easy_python_tests_build_info_fail
     Status: Success
     Size: 92.8KB | Time: 15.4s
     Summary: The build process failed during build info validation due to mismatched image repository and platform details, despite all Python unit tests passing. Two warnings were issued: one for unencrypted Docker password storage and another for cluster reachability issues. Immediate actions required include resolving build configuration mismatches, addressing network connectivity to the cluster, and improving credential storage security.
     Report: tmp/reports/report-2025-08-26T17-34-46.md

  ✅ easy_sharded_cluster_custom_podspec_max_api_keys
     Status: Success
     Size: 504.7KB | Time: 30.7s
     Summary: The analysis identified misconfigured RoleBindings, inconsistent image versions, and missing security contexts in CSI Kubernetes deployments, affecting 3 RoleBindings, 2 Roles, and 1 StatefulSet. These issues may lead to authorization errors, deployment instability, and security risks. Immediate actions required include verifying RoleBinding references, standardizing image versions, and auditing security contexts in all YAML manifests.
     Report: tmp/reports/report-2025-08-26T17-35-17.md

  ✅ easy_svet_compile_failed
     Status: Success
     Size: 605.1KB | Time: 26.7s
     Summary: A build failure occurred due to the use of an invalid compiler flag `/Wno-unused-parameter`, which is unsupported by MSVC and affected multiple source files. The error halted the build process entirely. Immediate action is required to review and correct the build configuration, ensuring compiler flags are set appropriately for MSVC, followed by retesting the build. Documentation of compiler-specific flag requirements is recommended to prevent recurrence.
     Report: tmp/reports/report-2025-08-26T17-35-44.md

  ✅ medium_build_init_database_base_image_not_found
     Status: Success
     Size: 90.6KB | Time: 18.2s
     Summary: The Docker image signing process failed due to a missing Amazon ECR repository (`julienben/staging-temp`), resulting in repeated task restarts and zero successful signings. Key impact includes aborted execution and wasted compute cycles. Critical actions needed are to verify the repository's existence, update configuration to reference a valid repository, and improve error handling and restart logic.
     Report: tmp/reports/report-2025-08-26T17-36-02.md

════════════════════════════════════════════════════════════
Done in 110.25s.
```

Note: processing time will strongly vary depending on configuration values, like chunking size, models used etc...

### Core Design Principles

1. **Separation of Concerns**: I/O operations (`dataLoader.ts`) are isolated from workflow logic
2. **Configuration-Driven**: All limits and models configurable via `config.ts` and env vars
3. **Fail-Fast**: Validation happens before expensive operations

### Key Components

```
src/mastra/workflows/
├── logCoreAnalyzerWorkflow.ts    # Main workflow orchestration
└── logCoreAnalyzer/
    ├── dataLoader.ts              # I/O with validation
    ├── config.ts                  # All configuration
    ├── constants.ts               # Type definitions
    ├── prompts.ts                 # LLM instructions
    └── __tests__/                 # Unit tests
```

### Adaptive Processing Strategy

```mermaid
graph TD
    B[Load & Validate] --> C{Size Check}
    C -->|Small| D[Single-Pass<br/>GPT-4.1]
    C -->|Large| E[Chunking]
    E --> F[Initial Analysis<br/>GPT-4.1]
    F --> G[Iterative Refinement<br/>GPT-4.1-nano]
    G --> H{More Chunks?}
    H -->|Yes| G
    H -->|No| I[Final Format<br/>GPT-4.1]
    D --> J[Output]
    I --> J
```

**Why this design?**
- Small files don't need chunking overhead
- Large files use cheaper models for middle chunks
- Maintains context through overlapping chunks

### Input
One of:
- **Local Files**: Direct file path analysis
- **URLs**: Remote file fetching with Evergreen API authentication
- **Raw Text**: Direct text input

And an optional `analysisContext`, which is passed through the whole process, to guide the agent on what to focus on, and provide additional context (such as name of the test, retries, name of the variant etc...)

### As a Workflow

```typescript
const result = await logCoreAnalyzerWorkflow.execute({
  url: 'https://example.com/logs/error.log',
  analysisContext: 'Focus on authentication failures'
});

// Returns:
// {
//   markdown: "# Technical Analysis Report...",
//   summary: "Brief executive summary..."
// }
```

### Benchmark Tool

```bash
# Analyze single file
yarn run-analyzer /path/to/file.log

# Batch processing with context
yarn run-analyzer --batch /logs/directory --context "Find memory leaks"

# Outputs performance metrics and saves reports to tmp/reports/
```

## Configuration

All limits are environment-configurable (see `config.ts` for defaults):

```env
LOG_ANALYZER_MAX_FILE_SIZE_MB=10
LOG_ANALYZER_MAX_TEXT_LENGTH=10000000
LOG_ANALYZER_MAX_TOKENS=200000
LOG_ANALYZER_URL_TIMEOUT_MS=30000
```

### Report

The workflow generates structured markdown reports with:
- **Overview**: Document summary
- **Key Entities**: Detected components/modules
- **Timeline**: Chronological events
- **Anomalies**: Errors and patterns
- **Metrics**: Quantitative analysis
- **Next Actions**: Recommendations
The report sections are defined in the prompts.

### Concise summary

On top of the report, a brief summary is generated, for quick overview, display in the console etc...

## Impact of the `analysisContext` feature

This feature enables to pass additional instructions to the agent, used during the whole summarization process.

This will be useful, to pass additional instructions that users request to the chat, but also to pass additional context about the logs (name of the task, project etc...).

Here is an example on file `easy_python_tests_build_info_fail`

Without additional context: 

`The build process failed during build info validation due to mismatched image repository and platform details, despite all Python unit tests passing. Two warnings were issued: one for unencrypted Docker password storage and another for cluster reachability issues. Immediate actions required include resolving build configuration mismatches, addressing network connectivity to the cluster, and improving credential storage security.`

When passing `"Please focus your analysis on understanding for which platforms the specific container images are built"` as additional context:

`Most container images are now built for both linux/amd64 and linux/arm64 platforms, expanding beyond the original linux/x86_64 target. However, some images like ops-manager remain limited to linux/amd64, causing test failures due to platform mismatches. Immediate action is needed to update test configurations for multi-architecture support and address single-platform build limitations.`

The analysis pivots sharply in the second paragraph, emphasizing content that’s minor in the original logs. When prompted, however, the agent correctly located and surfaced that information.

### Unit Tests for i/o operations
```bash
yarn test:run src/mastra/workflows/logCoreAnalyzer
```

Tests cover:
- Size/token limit validation
- URL fetching with timeouts
- Text normalization
- Error handling

### Performance Analysis

The benchmark tool (`run-analyzer`) provides:
- File processing times
- Success/failure rates  
- Performance bottleneck identification
- Detailed step-by-step analysis

## Integration

The workflow is provided to the main orchestrato, it can use it based on the description.

### Files Changed

**New Core Files:**
- `src/mastra/workflows/logCoreAnalyzerWorkflow.ts` 
- `src/mastra/workflows/logCoreAnalyzer/*.ts`

**New Tools:**
- `src/mastra/tests/scripts/run-log-analyzer.ts` - Benchmark CLI
- `src/mastra/tests/scripts/analyze-workflow-steps.ts` - Subfunctions for benchmark

**Integration:**
- Modified `src/mastra/index.ts`, `src/mastra/agents/evergreenAgent.ts`

## Suggested follow ups

- Auto compact of the summary during summarization loop to stay in the context limit.
- Fine tune prompts, hyperparameters (in `config.ts`) etc... based on our benchmark dataset.

### CODE REVIEW

File: src/mastra/workflows/logCoreAnalyzerWorkflow.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+const USER_INITIAL_PROMPT = (chunk: string, hint?: string) =>
+  `Summarize this first chunk into a concise technical brief.
+${hint ? `Context hint:\n${hint}\n` : ""}
+
+Chunk:
+"""${chunk}"""
+
+Return JSON:
+{ "updated": true, "summary": "<concise but comprehensive summary>", "evidence": ["<short quotes or line ranges>"] }`;
```
[the remark this answers is not in the document]
Comment (Julien-Ben) [reply src/mastra/workflows/logCoreAnalyzerWorkflow.ts:117]: Hey, that's already the case, you can check these lines:

`experimental_output: LoopStateOutputSchema`

"Experimental" name depends on which exact version of Mastra is used. This might change when I fix the dependencies. But it's the same parameter

**Review by khelif96 — COMMENTED** [review]
Great work! This works really well on the cases I tried
There are a few todos as part of this pr are you planning on addressing them here? If not could you file some tickets and link them in the TODOs so we can track them.

File: src/mastra/workflows/logCoreAnalyzerWorkflow.ts [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+  .dowhile(
+    refineStep,
+    async params =>
+      // Access inputData from the full params object
+      params.inputData.idx < params.inputData.chunks.length
+  )
+  .then(finalizeStep)
+  .commit();
+
+// This was initially a `.branch()` workflow step, but it involved too much complexity like unwrapping types correctly,
```
[the remark this answers is not in the document]
Comment (Julien-Ben) [reply src/mastra/workflows/logCoreAnalyzerWorkflow.ts:366]: I ran into many issues, it was a rabbit hole.

The way of Mastra of managing steps and workflow in branches was inconsistent with the documentation and varied between patches versions.

Typing issues were terrible to handle and debug.

When wrapping the workflow into a step, I also needed an "unwrap" step after the branch to make Mastra happy about typing. At this time I had formatting steps after the analysis steps.
So it means it will make it harder to add steps in the future if we keep the branching.

I also don't get the value of using branches for us right now. It looked good on paper but I don't understand what it gives us, except the display in the playground.

I leave it up to you to decide, but I feel it will bring way more issues than benefits to revert it to a branch based workfow.

File: src/mastra/workflows/logCoreAnalyzerWorkflow.ts [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+  // In the description, we tell the agent to use the "url" parameter, because that's the preferred way
+  // for the orchestrator to use this workflow, but all of them work.
+  description:
+    'Analyze, iteratively summarize, and produce a complete report of technical files of arbitrary types and structures. Pass the full URL of the file in the `url:` parameter.',
+  inputSchema: WorkflowInputSchema,
+  outputSchema: WorkflowOutputSchema,
+})
+  .then(loadDataStep) // Use the new unified load step with validation
+  .then(chunkStep)
+  .then(decideAndRunStep)
```
[the remark this answers is not in the document]
Comment (Julien-Ben) [reply src/mastra/workflows/logCoreAnalyzerWorkflow.ts:395]: Related to https://github.com/evergreen-ci/sage/pull/44#discussion_r2303488381

File: src/mastra/workflows/logCoreAnalyzer/config.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    ), // Max file size in MB
+    maxTextLength: parseInt(
+      process.env.LOG_ANALYZER_MAX_TEXT_LENGTH || '10000000',
+      10
+    ), // Max text length in characters (10M default)
+    maxUrlSizeMB: parseInt(
+      process.env.LOG_ANALYZER_MAX_URL_SIZE_MB || '10',
+      10
+    ), // Max size for URL fetches in MB
+    maxTokens: parseInt(process.env.LOG_ANALYZER_MAX_TOKENS || '2000000', 10), // Max estimated tokens to process
```
[the remark this answers is not in the document]
Comment (Julien-Ben) [reply src/mastra/workflows/logCoreAnalyzer/config.ts:42]: I'd say it's about the way we want to handle configuration in the project. For now it's easy to just manually set things in `config.ts`. For example when we fine tune the parameters based on our benchmarks.
But we can also configure everything via env variables. Up to us to decide

File: src/mastra/workflows/logCoreAnalyzer/prompts.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+export const MAX_FINAL_SUMMARY_TOKENS = 2048;
+
+// Agent Instructions
+export const INITIAL_ANALYZER_INSTRUCTIONS = `You are a senior engineer performing initial analysis of technical text (logs, code, configs, telemetry, build output).
+You always respond as compact JSON matching the provided schema.
+Focus on:
+- Understanding the overall structure and format of the content
+- Identifying key patterns, sections, and data types
+- Establishing context and technical domain
+- Preserving critical facts, identifiers, timestamps, error codes
```
[the remark this answers is not in the document]
Comment (khelif96) [reply src/mastra/workflows/logCoreAnalyzer/prompts.ts:14]: Can you please file a follow up ticket for this @Julien-Ben!

File: src/mastra/workflows/logCoreAnalyzerWorkflow.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (khelif96) [reply src/mastra/workflows/logCoreAnalyzerWorkflow.ts:366]: I think the biggest advantage to using the workflows is increased observability into what path a trace took. The type errors  seem to be related to https://github.com/evergreen-ci/sage/pull/44#discussion_r2301909373 this flag in the tsconfig causes typescript to be a little more strict in how it interprets types. Agent, Tools and workfows all type the `description` field differently.  `description?: string`  vs `description?: string | undefined` both of these typings technically do the same thing. but the `exactOptionalPropertyTypes` tsconfig flag forces typescript to think these are different types, Which is what causes the type errors you were seeing. I tried this out locally and managed to pass agents and workflows directly into the statements from a branch call without doing any aliasing.

**Review by khelif96 — APPROVED** [review]
Great job! Approving to unblock

File: src/mastra/workflows/logCoreAnalyzer/config.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Julien-Ben) [reply src/mastra/workflows/logCoreAnalyzer/config.ts:42]: Added this consideration in the follow up ticket
